src/lib/sm.h

```diff
@@ -0,0 +1,43 @@
+#ifndef __LIB_SM__
+#define __LIB_SM__
+
+#include <stdint.h>
+#include <stdbool.h>
+
+#define BITS(state) (1ULL << (state))
+
+enum {
+	SM_PREV_NONE  = -1,
+	/* sizeof(sm_conf::allowed * 8) */
+	SM_STATES_MAX = 64,
+	/* flags */
+	SM_INITIAL    = 1U << 0,
+	SM_FAILURE    = 1U << 1,
+	SM_FINAL      = 1U << 2,
+};
+
+struct sm_conf {
+	uint32_t    flags;
+	uint64_t    allowed;
+	const char *name;
+};
+
+struct sm {
+	int    rc;
+	int    state;
+	bool (*is_locked)(const struct sm *);
```

Comment:
so it looks like `is_locked` is optional, but `invariant` is required, maybe make that explicit somewhere in a comment.

Reply:
fixed.